rfc/src/rfcs/0003-resource-limits-timeouts.md

```diff
@@ -0,0 +1,224 @@
+Past tools for CBMC have required special annotations for "hard" harnesses to run serially.
+
+Further, to realize these benefits, we **must** implement resource limits by default.
+Consequently, this will likely be a "breaking change" for many customers.
+Either we should get feedback on whether this would be welcome, or we should consider a staged strategy for rolling out the feature (considered later in this doc).
+
+## User Experience
+
+The aim of this RFC is to introduce two new annotations to be placed along side proof harnesses.
+For example, from our unwinding tutorial:
+
+```rust
+#[cfg(kani)]
+#[kani::proof]
+#[kani::unwind(1)] // deliberately too low
+fn check_initialize_prefix()
+```
+
+This shows our abuse of `unwind` to force termination.
+We would prefer this to be unnecessary, and for the user to be introduced to unwinding via a default timeout.
+
+In the case where we do want to give higher limits, users should have to write something like:
+
+```rust
+#[cfg(kani)]
+#[kani::proof]
+#[kani::timeout(60)]
+#[kani::memory(10)]
+fn check_initialize_prefix()
+```
+
+> **Question 1: Should we consider alternative approach to specifying these limits?**
+>
+> I argue yes: We're now proliferating the number of attributes, so perhaps it's time to consider supporting:
+> `#[kani::proof(unwind = 5, timeout = 60, memory = 10)]`
+>
+> As we'll see later, the disadvantage of default limit choices is the friction of adding limit annotations to harnesses.
+> Making this as easy as possible helps mitigate those disadvantages.
+
+> **Question 2: Should we include units in the attributes?**
+>
+> I'm presently giving "seconds" and "gigabytes" and these seem totally reasonable units that we probably wouldn't ever deviate from.
+> However, it might be nice to either require units (to be explicit) or include units (if we think people might need them... perhaps just minutes).
+> I believe it would be reasonable to omit units however, and I'm not sure how difficult it would be to support `timeout = 2min` or `timeout = 4:00`.
+
+Beyond specifying limits on each proof harness, users would interact with these limits in two other ways:
+
+1. The output from running Kani.
+2. The command line arguments for manipulating limits.
+
+> **Question 3: How much flexibility should be provide on the command-line?**
+>
+> I believe we can keep this small.
+> I propose we only add a few new options, and I'm not sure if we should even go that far:
+> 1. `--default-timeout`
+> 2. `--default-memory`
+> 3. `--timeout-multiplier=X`
+> 4. `--ignore-limits`
+>
+> The purpose of the first two is not really for the command line but instead `Cargo.toml` (which currently is just an alternative way to supply command line arguments).
+>
+> The purpose of the multiplier is to allow a slow/inconsistent CI system to have an extra "grace period" for timeouts.
+> This way, CI can run `cargo kani --timeout-multiplier=1.3` to get 30% extra slack on timeouts.
```

Comment:
This one I'm not so convinced. I can see where you are coming from (our github workers are rather slow), but I would also advocate for looser timeouts.